fuel/datasets/base.py

```diff
@@ -177,88 +170,124 @@ def filter_sources(self, data):
+            self.iterables = [iterables[source] for source in self.sources]
+        else:
+            if not isinstance(iterables, collections.Iterable):
+                raise ValueError
+            self.iterables = [iterables]
+
+    def __len__(self):
+        return min(len(iterable) for iterable in self.iterables)
 
-        Any interaction with the outside world e.g. the file system,
-        database connections, servers, etc. should be done in this method.
-        This allows datasets to be pickled and unpickled, even in
-        environments where the original data is unavailable or has changed
-        position.
+    def open(self):
+        iterators = [_iter(channel) for channel in self.iterables]
+        return izip(*iterators)
+
+    def get_data(self, state=None, request=None):
+        if state is None or request is not None:
+            raise ValueError
+        return next(state)
 
-        """
-        pass
 
+class IndexableDataset(Dataset):
+    """Creates a dataset from a set of indexible containers.
 
+    Parameters
+    ----------
+    indexables : :class:`~collections.OrderedDict` or indexable
+        The indexable(s) to provide interface to. This means it must
+        support the syntax ```indexable[0]``. If an
+        :class:`~collections.OrderedDict` is given, its values should be
+        indexables providing data, and its keys strings that are used as
+        source names. If a single indexable is given, it will be given the
+        source ``data``.
+
+    Attributes
+    ----------
+    indexables : list
+        A list of indexable objects.
+
+    Notes
+    -----
+    If the indexable data is very large, you might want to consider using
+    the :func:`.do_not_pickle_attributes` decorator to make sure the data
+    doesn't get pickled with the dataset, but gets reloaded/recreated
+    instead.
+
+    This dataset also uses the source names to create properties that
+    provide easy access to the data.
 
+    """
+    def __init__(self, indexables, start=None, stop=None, **kwargs):
+        if isinstance(indexables, dict):
+            self.provides_sources = tuple(indexables.keys())
+        else:
+            self.provides_sources = ('data',)
+        super(IndexableDataset, self).__init__(**kwargs)
+        if isinstance(indexables, dict):
+            self.indexables = [indexables[source][start:stop]
+                               for source in self.sources]
+            if not all(len(indexable) == len(self.indexables[0])
+                       for indexable in self.indexables):
+                raise ValueError("sources have different lengths")
+        else:
+            self.indexables = [indexables]
+
+        def property_factory(source):
+            def property_(self):
+                return self.indexables[self.sources.index(source)]
+            return property_
+        for source in self.sources:
+            setattr(self.__class__, source, property(property_factory(source)))
```

Comment:
Thought about this, and it's actually not much of a problem. The worst thing that can happen is that it defines a property, and that some of the instances don't have a corresponding source. Currently that results in an `IndexError`, while it should really given an `AttributeError`, so I changed that.